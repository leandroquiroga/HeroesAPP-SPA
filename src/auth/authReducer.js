import { types } from "../types/type";



export const authReducer = (state = {}, action) => {
  const { type, payload } = action;
  
  switch (type) {
    case types.login:
      return {
        ...payload,
        logged: true,
      };
    case types.logout:
      return {
        logged: false,
      };
    
    default:
      return state;
  }
}