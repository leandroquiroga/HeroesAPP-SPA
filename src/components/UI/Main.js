import React from 'react';

export const Main = () => {
  const pathIMG = `/assets/img/not-super-hero.png`
  return (
       
    <div className=' container d-flex justyfy-content-center align-items-center'>

    {/*
        Si hay elemento en arreglo de favoritos,
        muestra una lista de todos los elementos 
    
     */}

    <div className='d-flex justify-content-between align-items-center w-100 p-5 flex-column flex-lg-row '>
      <img 
        className='img-fluid'
        alt='have not element'
        src={pathIMG}
      />

      <p className='text-center fs-3 mt-5 mt-lg-0'>
        No hay superheroes en la lista de favoritos...
      </p>
    </div>
  </div>
  );
};
