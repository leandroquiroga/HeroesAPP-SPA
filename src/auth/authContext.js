import { createContext, useEffect, useReducer, useState } from "react";
import { signInWithEmailAndPassword, signOut ,createUserWithEmailAndPassword } from 'firebase/auth';
import { authReducer } from "./authReducer";
import { auth } from './../firebase';
import { heroReducer } from "../helpers";

// Retorna el usuario si lo encuentra en localStorage si no retorna el logged en false
const init = () => JSON.parse(localStorage.getItem('user')) || {logged: false}

// Retorna los hero si lo encuntra en localStorage si no retorna un array 
const initHero = () => JSON.parse(localStorage.getItem('hero')) || [];

export const AuthContext = createContext();

export const AuthProvaider = (props) => {
  const [user, dispatch] = useReducer(authReducer, {}, init);
  const [currentUser, setCurrentUser] = useState({});
  const [hero, dispatchHero] = useReducer(heroReducer, [], initHero);

  // Cuando inicia la aplicacion va a estar observando si hay un cambio
  // de usuario.
  useEffect(() => {
    auth.onAuthStateChanged((user) => setCurrentUser(user))
  }, []);

  // Esta pendiente al user para ir almancendolo en el localStorage
  useEffect(() => {
    localStorage.setItem('user', JSON.stringify(user))
    return () => {
      if (!user) return;
    }
  }, [user]);

  // Esta pendiente al hero, cuando detecta un cambio lo almacena en el localStorage
  useEffect(() => {
    localStorage.setItem('hero', JSON.stringify(hero));
    return () => {
      if (!hero) return;
    }
  }, [hero])
  


  const singUp = (email, password) => createUserWithEmailAndPassword(auth, email, password);

  const logIn = (email, password) => signInWithEmailAndPassword(auth, email, password);
  
  const logOut = (auth) => signOut(auth);

  const value = {
    user,
    dispatch,
    singUp,
    logIn,
    logOut,
    currentUser,
    dispatchHero,
    hero,
  }
  return (
    <AuthContext.Provider value={value}>
      {props.children}
    </AuthContext.Provider>
  )
}