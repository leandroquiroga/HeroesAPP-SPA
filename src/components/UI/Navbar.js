import React, { useContext } from 'react';
import { Link, NavLink,  useNavigate } from 'react-router-dom';
import { AuthContext } from './../../auth/autoContext';
import { types } from './../../types/type';

export const NavbarScreen = () => {

  const navigate = useNavigate();
  const { user, dispatch } = useContext(AuthContext)
  // retorna a inicio de sesion

  const handleLoguot = () => {
    const action = {
      type: types.logout,
    };
  
    dispatch(action);
    navigate('/login', {
      replace: true
    })
  };
  return (
    <nav className='navbar navbar-expand-lg navbar-dark bg-dark px-2'>

      <Link
        className='navbar-brand'
        to="/"
      >
        Inicio
      </Link>
      <div className='navbar-collapse'>
        <div className='navbar-nav'>
          <NavLink
            className={({ isActive }) => 'nav-item nav-link text-white' + (isActive ? 'active' : '' )}
            to="/marvel"
          >
            Marvel
          </NavLink>

          <NavLink
            className={({ isActive }) => 'nav-item nav-link text-white' + (isActive ? 'active' : '' )}
            to='/dc'
          >
            DC
          </NavLink>
          <NavLink
            className={({ isActive }) => 'nav-item nav-link text-white' + (isActive ? 'active' : '' )}
            to='/search'
          >
            Search
          </NavLink>
      </div>
    </div>
    <div className='navbar-collapse collapse w-100 order-3 dual-collapse2 d-flex justify-content-end'>
      <ul className='navbar-nav ml-auto'>

        <span className='nav-item nav-link text-white'>
          Hola, {user.name} !
        </span>

        <button
          className='nav-item btn btn-danger'
          onClick={handleLoguot}
        >
          Cerrar sesi√≥n
        </button>
      </ul>
    </div>
  </nav>
  );
}