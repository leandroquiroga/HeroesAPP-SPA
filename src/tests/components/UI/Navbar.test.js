
import { mount } from 'enzyme';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import { AuthContext } from '../../../auth/authContext';
import { NavbarScreen } from '../../../components/UI/Navbar';


const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () =>  mockNavigate,
}));

describe('Testing on component <Navbar />', () => {

  const contexValue = {
    dispatch: jest.fn(),
    user: {
      name: 'Leandro',
      logged: true,
    },

  };
    
  const wapper = mount(
    <AuthContext.Provider value={contexValue}>
      <MemoryRouter initialEntries={['/']}>
        <Routes>
          <Route path='/' element={<NavbarScreen />} />
        </Routes>
      </MemoryRouter>
    </AuthContext.Provider>
  );

    
    
  test('This test must showing corretly', () => {
    const { user } = contexValue;
    const logoPath = '/assets/svg/avatar.svg';

    expect(wapper).toMatchSnapshot();
    expect(wapper.find('.nav-item.nav-link.text-white.text-center').text().trim()).toBe(`Hola, ${user.name} ðŸ‘‹`);
    expect(wapper.find('.navbar-icon').prop('src')).toBe(logoPath);
  });
});