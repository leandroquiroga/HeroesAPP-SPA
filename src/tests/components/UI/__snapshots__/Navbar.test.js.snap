// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing on component <Navbar /> This test must showing corretly 1`] = `
<nav
  className="navbar navbar-expand-lg navbar-dark bg-dark px-2"
>
  <section
    className="container-fluid"
  >
    <a
      className="navbar-brand"
      href="/"
      onClick={[Function]}
    >
      <img
        alt="logo navbar"
        className="img-fluid navbar-icon"
        src="/assets/svg/avatar.svg"
      />
    </a>
    <button
      className="navbar-toggler"
      onClick={[Function]}
      type="button"
    >
      <span
        className="navbar-toggler-icon"
      />
    </button>
    <article
      className="collapse navbar-collapse"
      id="navbarSupportedContent"
    >
      <div
        className="navbar-nav text-center "
      >
        <a
          className="nav-item nav-link text-white"
          href="/marvel"
          onClick={[Function]}
        >
          Marvel
        </a>
        <a
          className="nav-item nav-link text-white"
          href="/dc"
          onClick={[Function]}
        >
          DC
        </a>
        <a
          className="nav-item nav-link text-white"
          href="/search"
          onClick={[Function]}
        >
          Search
        </a>
      </div>
      <div
        className="navbar-nav ms-auto"
      >
        <ul
          className="navbar-nav"
        >
          <span
            className="nav-item nav-link text-white text-center"
          >
            Hola, 
            Leandro
            <small
              style={
                Object {
                  "fontSize": "20px",
                }
              }
            >
               ðŸ‘‹ 
            </small>
          </span>
          <button
            className="nav-item btn btn-danger"
            onClick={[Function]}
          >
            Cerrar sesiÃ³n 
            <img
              alt="logout"
              src="/assets/svg/logout.svg"
            />
          </button>
        </ul>
      </div>
    </article>
  </section>
</nav>
`;
